### ISSUE TITLE
Purge sometimes removes wrong files

### ISSUE THREAD

**Issue by warwickmm** [duplicati/duplicati#3927]
## Environment info
- **Duplicati version**: Reproduced on 2.0.4.5 and 2.0.4.30
- **Operating system**: Linux (cannot reproduce on Windows)

## Description
The [documentation](https://duplicati.readthedocs.io/en/latest/04-using-duplicati-from-the-command-line/#the-purge-command) seems to imply that the files provided to `purge` (either by a list or via `--include=<pattern>`) are the files to be removed.  However, sometimes they are the files that are _kept_, and _all non-matching_ files are removed.

## Steps to reproduce

1. Create a backup of a single source folder that contains 3 files: `a`, `b`, and `c`.
2. Run a single backup.
3. In the UI, go to "Commandline...", and select the "purge" command.
4. In the "Commandline arguments" field, enter
```
<path to source>/a
```
or
```
--include=<path to source>/a
```
5. Click on "Run "purge" command now".

The expectation is that file `a` is removed from the backup sets.  However, the opposite has occurred.  Files `b` and `c` are removed.  This can be seen in the resulting list of files in the restore tree, or by viewing the `DryRun` logs (include `--dry-run` in the command line arguments in step 4 above):
```
Sep 28, 2019 12:57 PM: Would upload file duplicati-20190928T195632Z.dlist.zip.aes (230 bytes) and delete file duplicati-20190928T195631Z.dlist.zip.aes, removing 3 files
Sep 28, 2019 12:57 PM: Would write files to remote storage
Sep 28, 2019 12:57 PM: Purging file /home/ken/no_backup/duplicati_tests/source/c (0 bytes)
Sep 28, 2019 12:57 PM: Purging file /home/ken/no_backup/duplicati_tests/source/b (0 bytes)
```

If in step 4 we specify
```
*/a
```
or
```
--include=*/a
```
we see that only file `a` is removed:
```
Sep 28, 2019 1:03 PM: Would upload file duplicati-20190928T200335Z.dlist.zip.aes (230 bytes) and delete file duplicati-20190928T200334Z.dlist.zip.aes, removing 1 files
Sep 28, 2019 1:03 PM: Would write files to remote storage
Sep 28, 2019 1:03 PM: Purging file /home/ken/no_backup/duplicati_tests/source/a (0 bytes)
```

This can also be reproduced from the command line
```
#!/bin/bash

source=$(mktemp -d)
destination=$(mktemp -d)
dbpath=$(mktemp)

# Create source files
touch "${source}"/a
touch "${source}"/b
touch "${source}"/c

# Backup
printf "\nRunning backup...\n\n"
mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe backup "file:///${destination}" "${source}" --dbpath="${dbpath}" --passphrase="password"

# List files
printf "\nListing files...\n\n"
mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe find "file:///${destination}" "${source}" --dbpath="${dbpath}" --include="*" --passphrase="password"

# Purge one file
printf "\nPurging file...\n\n"
mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe purge "file:///${destination}" "${source}/a" --dbpath="${dbpath}" --passphrase="password"

# List files
printf "\nListing files...\n\n"
mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe find "file:///${destination}" "${source}" --dbpath="${dbpath}" --include="*" --passphrase="password"
```
yields
```
Running backup...

Backup started at 9/28/2019 1:16:13 PM
Checking remote backup ...
  Listing remote folder ...
Scanning local files ...
  4 files need to be examined (0 bytes)
  Uploading file (1.11 KB) ...
  Uploading file (925 bytes) ...
  Uploading file (1021 bytes) ...
Compacting remote backup ...
Checking remote backup ...
  Listing remote folder ...
Verifying remote backup ...
Remote backup verification completed
  Downloading file (1021 bytes) ...
  Downloading file (925 bytes) ...
  Downloading file (1.11 KB) ...
  0 files need to be examined (0 bytes)
  Duration of backup: 00:00:01
  Remote files: 3
  Remote size: 3.01 KB
  Total remote quota: 19.56 GB
  Available remote quota: 6.34 GB
  Files added: 3
  Files deleted: 0
  Files changed: 0
  Data uploaded: 3.01 KB
  Data downloaded: 3.01 KB
Backup completed successfully!

Listing files...

Listing contents 0 (9/28/2019 1:16:13 PM):
/tmp/tmp.D9KiK4TjGs/ 
/tmp/tmp.D9KiK4TjGs/a (0 bytes)
/tmp/tmp.D9KiK4TjGs/b (0 bytes)
/tmp/tmp.D9KiK4TjGs/c (0 bytes)

Purging file...

  Listing remote folder ...
  Uploading file (829 bytes) ...
  Deleting file duplicati-20190928T201613Z.dlist.zip.aes ...

Listing files...

Listing contents 0 (9/28/2019 1:16:14 PM):
/tmp/tmp.D9KiK4TjGs/a (0 bytes)
```
If we instead ran the following purge command
```
mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe purge "file:///${destination}" "*/a" --dbpath="${dbpath}" --passphrase="password"
```
the file list shows
```
Listing contents 0 (9/28/2019 1:18:12 PM):
/tmp/tmp.RtNs17dNo8/ 
/tmp/tmp.RtNs17dNo8/b (0 bytes)
/tmp/tmp.RtNs17dNo8/c (0 bytes)
```

## TLDR
If the files are specified with a wildcard, the specified files are removed.  If the names of the files are provided, _all other files are removed_.

**Comment by ts678** [comment]
What version is this in (and is there any known version it's not in). Works OK for me on 2.0.4.28 canary and 2.0.4.30 canary, although `find` with just source directory found nothing. Needed to add * on end. Dlist file was corrupted down to just `manifest` file on 2.0.4.30, but that's expected given known issue:
[Purge operations result in broken dlist files #3924](https://github.com/duplicati/duplicati/issues/3924)

**Comment by warwickmm** [comment]
This fails for me on 2.0.4.28.  Are you using wildcards in your include pattern or the full path to the file?  For me, the existence of `*` results in the expected behavior.  Using a full path to specify the file results in the unexpected behavior.

I'll try and bisect this.  A quick glance at the relevant code seems to indicate that this might have been the case since 2016...

**Comment by warwickmm** [comment]
Fails for me on 2.0.4.5 as well.
```
$ mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe backup file:////tmp/tmp.uBycxZ8w0C /tmp/tmp.VnigedfVjM --dbpath=/tmp/tmp.AZutP5r5Fk --passphrase=password > /dev/null
$ mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe purge file:////tmp/tmp.uBycxZ8w0C /tmp/tmp.VnigedfVjM/a --dbpath=/tmp/tmp.AZutP5r5Fk --passphrase=password > /dev/null
$ mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe find file:////tmp/tmp.uBycxZ8w0C /tmp/tmp.VnigedfVjM --dbpath=/tmp/tmp.AZutP5r5Fk --include=* --passphrase=password
Listing contents 0 (9/28/2019 6:26:19 PM):
/tmp/tmp.VnigedfVjM/a (0 bytes)
```
Note that `<path to source>/a` was provided as the file to purge above.  After the purge, only file `a` remains.

If I modify the purge pattern by appending `*` to the end, the results are different.  Files `b` and `c` remain after the purge:
```
$ mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe backup file:////tmp/tmp.iBKjWm8zMo /tmp/tmp.Lr8aUd7Bcj --dbpath=/tmp/tmp.7qY27e2e10 --passphrase=password > /dev/null
$ mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe purge file:////tmp/tmp.iBKjWm8zMo "/tmp/tmp.Lr8aUd7Bcj/a*" --dbpath=/tmp/tmp.7qY27e2e10 --passphrase=password > /dev/null
$ mono Duplicati/GUI/Duplicati.GUI.TrayIcon/bin/Debug/Duplicati.CommandLine.exe find file:////tmp/tmp.iBKjWm8zMo /tmp/tmp.Lr8aUd7Bcj --dbpath=/tmp/tmp.7qY27e2e10 --include=* --passphrase=password
Listing contents 0 (9/28/2019 6:31:30 PM):
/tmp/tmp.Lr8aUd7Bcj/ 
/tmp/tmp.Lr8aUd7Bcj/b (0 bytes)
/tmp/tmp.Lr8aUd7Bcj/c (0 bytes)

```

**Comment by warwickmm** [comment]
Looks like all the unit tests happen to have a `*` in the filter pattern.

**Comment by ts678** [comment]
>  Are you using wildcards in your include pattern or the full path to the file?

Paths, because that was said to cause the issue. Example below deletes A.txt, leaving B.txt and C.txt:

> C:\>"C:\ProgramData\Duplicati\duplicati-2.0.4.28_canary_2019-09-05\Duplicati.CommandLine.exe" purge "file://C:\ProgramData\Duplicati\duplicati-2.0.4.28_canary_2019-09-05\RUN\purge test 1\\" "C:\backup source\A.txt" --dry-run=true --backup-name="purge test 1" --dbpath="C:\ProgramData\Duplicati\duplicati-2.0.4.28_canary_2019-09-05\RUN\WZKDMCRLPS.sqlite" --encryption-module= --compression-module=zip --dblock-size=50mb --no-encryption=true --no-auto-compact=true --log-file-log-level=Profiling --profile-all-database-queries=true --log-file="C:\ProgramData\Duplicati\duplicati-2.0.4.28_canary_2019-09-05\\RUN\purge test 1.Profiling.log" --disable-module=console-password-input --console-log-level=Information
The operation PurgeFiles has started
Starting purge operation
Backend event: List - Started:  ()
  Listing remote folder ...
Backend event: List - Completed:  (3 bytes)
Replacing fileset duplicati-20190929T000243Z.dlist.zip with duplicati-20190929T000244Z.dlist.zip which has with 1 fewer file(s) (1 bytes reduction)
[Dryrun]:   Purging file C:\backup source\A.txt (1 bytes)
[Dryrun]: Would write files to remote storage
[Dryrun]: Would upload file duplicati-20190929T000244Z.dlist.zip (335 bytes) and delete file duplicati-20190929T000243Z.dlist.zip, removing 1 files

**Comment by warwickmm** [comment]
Interesting...I wonder if the results are different on Windows vs. Linux.  If I modify [this line](https://github.com/duplicati/duplicati/blob/c412bc9cae0db9780f7c219bed34c17bef5ef1c8/Duplicati/UnitTest/PurgeTesting.cs#L95) in the unit tests from
```
var res = c.PurgeFiles(new Library.Utility.FilterExpression("*" + Path.DirectorySeparatorChar + single_version_candidate));
```
to
```
var res = c.PurgeFiles(new Library.Utility.FilterExpression(Path.Combine(this.DATAFOLDER, single_version_candidate)));
```
then the test fails with
```
  Incorrect number of removed files after single-versions purge
  Expected: 1
  But was:  15
```

I'll run this modified test on AppVeyor and Travis to see if the results indicate some platform-dependent behavior.

**Comment by warwickmm** [comment]
Ok, the modified test passed in AppVeyor and failed in Travis.  This seems to affect Linux and not Windows.  Thanks @ts678 for helping test.

**Comment by warwickmm** [comment]
It turns out there's a separate code path for case-sensitive filesystems.  The query in that case needed a small modification.  I'll post the fix after some more testing.

**Comment by warwickmm** [comment]
Fixed in pull request #3929.
